include/circt/Dialect/HW/HWModuleGraph.h

```diff
@@ -0,0 +1,189 @@
+        .Case<circt::comb::ShrSOp, circt::comb::ShrUOp>(
+            [&](auto) { return ">>"; })
+        .Case<circt::comb::ShlOp>([&](auto) { return "<<"; })
+        .Case<circt::comb::ICmpOp>([&](auto op) {
+          switch (op.getPredicate()) {
+          case circt::comb::ICmpPredicate::eq:
+          case circt::comb::ICmpPredicate::ceq:
+          case circt::comb::ICmpPredicate::weq:
+            return "==";
+          case circt::comb::ICmpPredicate::wne:
+          case circt::comb::ICmpPredicate::cne:
+          case circt::comb::ICmpPredicate::ne:
+            return "!=";
+          case circt::comb::ICmpPredicate::uge:
+          case circt::comb::ICmpPredicate::sge:
+            return ">=";
+          case circt::comb::ICmpPredicate::ugt:
+          case circt::comb::ICmpPredicate::sgt:
+            return ">";
+          case circt::comb::ICmpPredicate::ule:
+          case circt::comb::ICmpPredicate::sle:
+            return "<=";
+          case circt::comb::ICmpPredicate::ult:
+          case circt::comb::ICmpPredicate::slt:
+            return "<";
+          }
+          assert(false && "unhandled ICmp predicate");
+        })
+        .Case<circt::seq::CompRegOp>([&](auto) { return "reg"; })
+        .Case<circt::hw::ConstantOp>([&](auto op) {
+          llvm::SmallString<64> valueString;
+          op.getValue().toString(valueString, 10, false);
+          return valueString.str().str();
+        })
+        .Default([&](auto op) { return op->getName().getStringRef().str(); });
+  }
+
+  std::string getNodeAttributes(circt::hw::detail::HWOperation *node,
+                                circt::hw::HWModuleOp) {
+    std::string attrs;
+
+    llvm::TypeSwitch<mlir::Operation *>(node)
+        .Case<circt::hw::ConstantOp>(
+            [&](auto) { attrs = "fillcolor=darkgoldenrod1,style=filled"; })
+        .Case<circt::comb::MuxOp>([&](auto) {
+          attrs = "shape=invtrapezium,fillcolor=bisque,style=filled";
+        })
+        .Case<circt::hw::OutputOp>(
+            [&](auto) { attrs = "fillcolor=lightblue,style=filled"; })
+        .Default([&](auto op) {
+          llvm::TypeSwitch<mlir::Dialect *>(op->getDialect())
+              .Case<circt::comb::CombDialect>([&](auto) {
+                attrs = "shape=oval,fillcolor=bisque,style=filled";
+              })
+              .template Case<circt::seq::SeqDialect>([&](auto) {
+                attrs = "shape=folder,fillcolor=gainsboro,style=filled";
+              })
+              .Default([&](auto) {});
+        });
+    return attrs;
+  }
+
+  static void
+  addCustomGraphFeatures(circt::hw::HWModuleOp mod,
+                         llvm::GraphWriter<circt::hw::HWModuleOp> &g) {
+
+    // Add module input args.
+    auto &os = g.getOStream();
+    os << "subgraph cluster_entry_args {\n";
+    os << "label=\"Input arguments\";\n";
+    for (auto arg : mod.getPorts().inputs) {
+      g.emitSimpleNode(reinterpret_cast<void *>(arg.getId()), "",
```

Comment:
Do we need to worry about IDs aliasing with others generated internally to `g?

Reply:
Like, should we actually be using a pointer?

Reply:
Ughh... welcome to graphtraits. Nodes are printed as follows:

```c++
void emitSimpleNode(const void *ID, ...) {
   O << "\tNode" << ID ...
}
```
All other nodes will thus have some heap-allocated pointer prefix as their identifier in the `dot` file. So to adhere to this "convention", we could either heap-allocate these arg IDs and pass that pointer, or just short circuit (like above) and assume that other nodes in the graph wont allocate at addresses 0x0, 0x1, ... (which i think is a totally fair assumption).

Reply:
We could also use the actual pointers. Like the memory addresses into the list of ports in the IR stored in the heap?

Reply:
Just had a stab at this - for some reason, even when ensuring that references are used, `mod.getPorts().inputs` iteration returns stack-allocated copies and thus are not stable.